File: prog4.go
package main

import (
	"fmt"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	var n, m int
	fmt.Scan(&n, &m)

	arr := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}

	pavelScore, vikaScore := 0, 0
	idx := 0
	prevK := 0
	pavelTurn := true

	for idx < n {
		maxSum := -1000000000 // Аналог минус бесконечности
		bestK := 0
		limit := min(m, n-idx)

		for k := 1; k <= limit; k++ {
			if k == prevK {
				continue
			}

			currentSum := 0
			for i := 0; i < k; i++ {
				currentSum += arr[idx+i]
			}

			if currentSum > maxSum {
				maxSum = currentSum
				bestK = k
			}
		}

		if bestK == 0 {
			break
		}

		if pavelTurn {
			pavelScore += maxSum
		} else {
			vikaScore += maxSum
		}

		idx += bestK
		prevK = bestK
		pavelTurn = !pavelTurn
	}

	if pavelScore > vikaScore {
		fmt.Println(1)
	} else {
		fmt.Println(0)
	}
}

File: prog4.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int n, m;
    if (!(cin >> n >> m)) return 0;

    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];

    int pavel_score = 0, vika_score = 0;
    int idx = 0;
    int prev_k = 0;
    bool pavel_turn = true;

    while (idx < n) {
        long long max_sum = -1e18; // очень маленькое число
        int best_k = 0;
        int limit = min(m, n - idx);

        for (int k = 1; k <= limit; k++) {
            if (k == prev_k) continue; // Правило 2 варианта

            long long current_sum = 0;
            for (int i = 0; i < k; i++) {
                current_sum += arr[idx + i];
            }

            if (current_sum > max_sum) {
                max_sum = current_sum;
                best_k = k;
            }
        }

        // Если ходов нет из-за ограничений, вынуждены прервать или пропустить
        if (best_k == 0) break;

        if (pavel_turn) pavel_score += max_sum;
        else vika_score += max_sum;

        idx += best_k;
        prev_k = best_k;
        pavel_turn = !pavel_turn;
    }

    if (pavel_score > vika_score) cout << 1 << endl;
    else cout << 0 << endl;

    return 0;
}
